### PR TITLE
Mieubrisse/readme upgrade

### PR DESCRIPTION
**Pull request by mieubrisse**
oH bAbY

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,69 +1,52 @@
...
+* [Examples](#examples)
+* [Notes](#notes)
+    * [Abnormal Exit](#abnormal-exit)
+    * [Container, Volume, &amp; Image Tidying](#container-volume--image-tidying)
+
+Architecture
+------------
+The Kurtosis architecture has four components:
+
+1. The **test network**, composed of Docker containers running the services necessary for a given test
```
Comment (galenmarchetti) [inline README.md:16]: I think the phrasing of this is a bit confusing given that for a given repository implementing kurtosis tests there will be multiple test networks (there is no THE test network) and multiple controllers as wel

Comment (galenmarchetti) [reply README.md:16]: maybe just making test network and controllers plural, to emphasize to the reader that there are multiple instances of each running for a given CI cycle

File: TUTORIAL.md [hunk 2]
```diff
@@ -0,0 +1,368 @@
+# Kurtosis Implementation Tutorial
+To run tests with Kurtosis, we'll need to define custom components for producing each of the four Kurtosis components. At a high level, this means writing:
+
+1. A **network** definition that defines a cluster of services that a test will use
```
Comment (galenmarchetti) [inline TUTORIAL.md:4]: again worth rewording this to reflect that there will be multiple networks for one test suite

Comment (galenmarchetti) [reply TUTORIAL.md:4]: could just be "network definitions that define a cluster of services that each test will use"

File: TUTORIAL.md [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+1. A **test suite** packaging all the tests for your application
+1. A **controller** Docker image that runs code wrapping Kurtosis' controller library
+1. An **initializer** CLI that wraps Kurtosis' initializer library
+
+In the below tutorial we'll show how to implement each of these components from scratch!
+
+## The Test Network
+A test always runs against a network of services being tested, and we'll first need to tell Kurtosis what that network should look like. Networks are composed of services, so our firs step is defining what a "service" looks like.
```
Comment (galenmarchetti) [inline TUTORIAL.md:12]: typo in this line in "first": "firs" -> "first"

File: TUTORIAL.md [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    // We'll imagine the Docker image running our service is hardcoded to listen on port 80
+    httpRestPort nat.Port = "80/tcp"
+
+    configFileKey = "config-file"
+)
+
+type MyServiceInitializerCore struct {
+    // This is a parameter specific to our service
+    // We could put more service-specific parameters here if needed
+    MyServiceParam1 string
```
Comment (galenmarchetti) [inline TUTORIAL.md:57]: IMO worth naming this something more concrete thats readable in the example, for example "MyServiceCLIParam1" or something. just so people can fit something in their brains for what kind of things might be considered parameters for a ServiceInitializerCore

File: TUTORIAL.md [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    // Return a filepath on our Docker image that's safe to mount the Kurtosis test volume on
+}
+
+func (core MyServiceInitializerCore) GetStartCommand(mountedFileFilepaths map[string]string, publicIpAddr string, dependencies []Service) ([]string, error) {
+    // Use our specific knowledge of the Docker image to craft the command the Docker image will run with
+    // This is where we'll use any service-specific params from above, including MyServiceParam1
+}
+```
+
+Note the service "dependencies" that show up above. Kurtosis knows that some services will depend on others, and gives the developer the option to modify a service's files and start command based on other existing services in the network. We'll see how to declare these dependencies later.
```
Comment (galenmarchetti) [inline TUTORIAL.md:90]: +100 for including this description

File: TUTORIAL.md [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+}
+```
+
+We're all set up to use this service in a network now... but of course, we still need to define what that network looks like. Kurtosis has a [ServiceNetwork](https://github.com/kurtosis-tech/kurtosis/blob/develop/commons/networks/service_network.go) object that represents the underlying state of the test network, but interacting with it is often too low-level for writing clean tests. To make writing tests as simple as possible, Kurtosis lets the developer define an arbitrary network struct that wraps the low-level network representation; this struct will then be passed to the tests. The developer can define this higher-level network wrapper object any way they please, but in our example we'll imagine that our tests all use a three-node network. Thus, our wrapper struct looks like so:
+
+```go
+type ThreeNodeNetwork struct {
+    networks.Network
+
+    BootNode MyService
```
Comment (galenmarchetti) [inline TUTORIAL.md:122]: can we rename these "PeerNode1" and "PeerNode2" ? Not sure which representative systems might model services in a parent-child relationship, boot-peer might be more tangible

**Review by galenmarchetti — APPROVED** [review]
I left some minor comments - approved pending revision of those.

On a larger note, I _think_ the tutorial might be better understandable if we had an actually system (like a bitcoin or a DApp test, or something). There's a bit of a mental jump that has to happen to imagine what kinds of services, start commands, parameters, networks might exist given the highly abstract tutorial we have now...but let's roll with this because its very thorough, and if we see people having difficulty then later we can try to provide a concrete one (which will definitely take more work / more verification etc)

File: TUTORIAL.md (hunk 4 above)
Comment (mieubrisse) [reply TUTORIAL.md:57]: Changed to `MyServiceLogLevel`

File: TUTORIAL.md (hunk 6 above)
Comment (mieubrisse) [reply TUTORIAL.md:122]: Renamed to `DependentNode{1,2}` to make even clearer (PeerNode to me implies equality, which IMO isn't true in this case as these two guys depend on the boot node)

**Pull request merged**